### PR TITLE
Add test reporting, test suite definitions, and test notifier

### PR DESCRIPTION
**Pull request by TGPSKI**
This PR refactors Cavy to add test suite definitions, test reporting, and test notification. Major changes were made to run() in TestScope.js, with minor supporting changes to other components. In addition, the ./cavy/reporting directory is added, supporting test report generation in supported formats. Currently, Chromium JSON Test Output is supported, with an example test reporter class included for others to contribute new reporting handlers. 

The sample project has been updated to reflect the refactoring as well. The sample project includes the recording server, as well as a webhook testing feature. 

Other major changes to the sample project include:
- Conditional test execution (./helpers/cavy.js)
- Example of testWrapHook by making EmployeeListItem stateless
- Added eslint config
- Added test server & support code

Please let me know what questions and comments you have!

Thanks, Tyler

*BTW - this is an amazing package, and we use this in production. We've been looking for something like this for quite some time now.*

Sample test output: https://github.com/TGPSKI/cavy/blob/add-jenkins-reporting/sample-app/EmployeeDirectory/specs/server/output/testResults_2017-03-14_11_42_19.json

Example video: https://drive.google.com/file/d/0B0idOjd0AFReZy1yRXRLQUtXb00/view?usp=sharing

image

# Tester.js
Pass new props to TestScope.js
```
  this.props.testStartDelay, 
  this.props.consoleLog, 
  this.props.reporter, 
  this.props.notifier
```

# TestScope.js

## constructor
Add new props passed in via props from <Tester>

```
  testStartDelay:    int, optional.              - pause before test execution begins (ms)
  consoleLog:        tristate, optional          - determine level of console feedback
                     false: no console.log statements
                     true: some console.log statements
                     verbose: detailed console.log statements

  reporter:          string, optional            - select reporter formatting for test report output
  notifier:          obj, optional               - pass notifier parameters to call webhook on test completion
```

## run
- Add test battery start / stop timestamp
- Add optional testStartDelay 
- Loop through testSuites
  - log suite start/stop timestamp
  - Loop through testCase
    - log case start/stop timestamp
    - optional: report result to console.log/warn
    - remove f from test case to simplify reporting
    - save test case results to class object
- On test battery complete
  - optional: add all results to reporter
  - optional: reporter generates test report
  - optional: notify webhook of test completion

## _handleConsoleLog(log, verbose, warn)
- Handles reporting to console based on TestScope props 
- Handles different logging levels
- Handles console.log vs console.warn on failed tests

## _handleNotify(testReport)
- If notifier enabled through prop, posts test results to API

# Project structure
```
~/foo/mySampleApp                         -> react native root
           ./specs                        -> cavy test folder
                ./itActions.js            -> individual test specs
                ./itSuites.js             -> test suites built from combining test specs
                ./itTestComponents.js     -> cavy-only test components that bypass
                ./doc
                    ./itElements.md       -> list of all testable elements
                    ./itActionsList.md    -> list of all exported funcitons in itActions.js
                ./server


```
## index.ios.js

Shows how to handle turning tests on and off
Shows new prop definitions
Shows test suite imports
```
import React, { PureComponent } from 'react';
import {AppRegistry} from 'react-native';

import { Tester, TestHookStore } from 'cavy';

import EmployeeDirectoryApp from './app/EmployeeDirectoryApp';

import GLOBAL from './app/helpers/globals.js';

if (GLOBAL.TEST_ENABLED) {
  var testHookStore = new TestHookStore();

  var TestSuites = require('./specs/itSuites.js');
  var TestSuitesArray = [TestSuites.verifyAndSearch, TestSuites.verifyActionBarEmail, TestSuites.secretPresence, TestSuites.runFailures];
 
  var testApiParams = {
    url: 'http://localhost:3003/jenkins',
    method: 'POST',
    headers: {'Content-Type': 'application/json'}
  };
}

class AppWrapper extends PureComponent {
  render() {
    if (GLOBAL.TEST_ENABLED) {
      return (
        <Tester 
          specs={TestSuitesArray} 
          store={testHookStore} 
          waitTime={1000}
          testStartDelay={1000}
          consoleLog='verbose' //{false}, {true}, 'verbose'
          reporter='ChromiumJSONTestReporting'
          notifier={testApiParams}
          reRender={true}
        >
        <EmployeeDirectoryApp />
      </Tester>
      );
    } else {
      return (
      <EmployeeDirectoryApp />
      );
    }
  }
}

AppRegistry.registerComponent('EmployeeDirectory', () => AppWrapper);
```

## Cavy.js

Cavy.js allows you to generate test hooks when testing is enabled, and removes the code overhead in production.

```
////// cavy //////
import { hook, wrap } from 'cavy';

import GLOBAL from './Globals';

export function testWrapHook(component) {
  if (GLOBAL.TEST_ENABLED) {
    return hook(wrap(component));
  } else {
    return component;
  }
}

export function testHook(component) {
  if (GLOBAL.TEST_ENABLED) {
    return hook(component);
  } else {
    return component;
  }
}
```

## CI Integration
```
- Trigger build on Jenkins
- Jenkins builds iOS / Android -> Parameterized build which sets ./helpers/globals.js GLOBAL.TEST_ENABLED = true;
- Jeknins runs `npm run server`
- Post-build trigger to install & run build products on simulators or connected devices
- Test execution
- Cavy reporters send formatted report to test report server
- Trigger webhook API request to Jenkins
- Jenkins API webhook triggers new job which saves test output to jenkins portal
- If test === pass, then run parameterized build which sets ./helpers/globals.js GLOBAL.TEST_ENABLED = false;
```

### CODE REVIEW

**Comment by jagranata** [comment]
:+1:

**Comment by Fathalian** [comment]
👍

**Comment by saivann** [comment]
:+1:

**Comment by samarchyan** [comment]
👍

**Comment by TGPSKI** [comment]
Thanks to @j2kun for the updates to `server/controllers/jenkins.js`, now starting to support xUnit XML reporting.

Changes since the PR opened:
- Refactor _handleConsoleLog to remove use of ternary operator without assignment
- Update sample-app
             - new secretSearch test hook
             - updated example in index.ios.js for conditional test running (i.e. don't run tests if API_URL is pointed to prod

**Comment by TGPSKI** [comment]
Jenkins <3 Cavy

image

**Comment by jalada** [comment]
Hey @TGPSKI et al, thanks for your continued work :)

This PR is getting pretty hefty though, how might we split it up into different parts to get more of a grip on it?

**Comment by TGPSKI** [comment]
@jalada: sure thing. I think this can be broken into 2. The change log looks a little overwhelming because I did a lot of stuff to the sample project.

1) Add test suites, timestamping, and reporting to TestScope.js
2) Add the sample project with stateless component support, turing tests off in prod, and the notify server

File: src/Tester.js [hunk 1]
```diff
@@ -98,14 +107,20 @@ Tester.propTypes = {
   store: PropTypes.instanceOf(TestHookStore),
   specs: PropTypes.arrayOf(PropTypes.func),
   waitTime: PropTypes.number,
-  clearAsyncStorage: PropTypes.bool
+  clearAsyncStorage: PropTypes.bool,
+  testStartDelay: PropTypes.number,
+  consoleLog: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
+  reporter: PropTypes.string,
```
Comment (jalada) [inline src/Tester.js]: Could you not get the user to pass in the actual class, instead of a string representing it?

File: src/Tester.js [hunk 2]
```diff
@@ -63,7 +72,7 @@ export default class Tester extends Component {
   }
 
   async runTests() {
-    scope = new TestScope(this, this.props.waitTime);
+    let scope = new TestScope(this, this.props.waitTime, this.props.testStartDelay, this.props.consoleLog, this.props.reporter, this.props.notifier, this.props.reRender);
```
Comment (jalada) [inline src/Tester.js]: Adding a whole bunch of sliced out props as arguments here feels a bit nasty. Would an object be a better API?

**Comment by jalada** [comment]
Thanks @TGPSKI.

Out of interest, what was the motivation for adding the concept of suites to test specs? Is it for better parity with Jenkins, or your alternate reporters, or something like that?

File: src/Tester.js (hunk 1 above)
Comment (TGPSKI) [inline src/Tester.js]: 👍

File: src/Tester.js (hunk 2 above)
Comment (TGPSKI) [inline src/Tester.js]: 👍

**Comment by TGPSKI** [comment]
I found a need for suites when constructing my test actions. I see specs as small functional components you compile into test suites, and may re-use multiple times. For example, my spec definition file used at work contains five categories of specs. These all define basic interactions with the app, like navigating to a signup page, inputting text in a form, or ensuring that an error message doesn't show after an input action.

* Secret Hooks
* Navigation
* Input / Input + Submit
* Press
* Presence / Not Presence

Test suites are collections of test specs that allow you to re-use lowest level testing functionality as needed.

I'll use two examples of our test suites to explain further. We test a variety of error and success conditions in our app as part of integration testing. One basic thing to test is user signup success and failure.

## test suite #1: emailSignupError

- spec #1: navSignupEmailSignup
- spec #2: inputEmailSignupSubmit(spec, email, username, password)
- spec #3: presenceEmailSignupError
- spec #4: navEmailSignupSignup

## test suite #2: emailSignupSuccess

- spec #1: navSignupEmailSignup
- spec #2: inputEmailSignupSubmit(spec, email2, username2, password2)
- spec #3: notPresenceEmailSignupError
- spec #4: navEmailSignupInbox

By separating lower level actions from compiled suites, we are able to re-use specs throughout app tests. Making the inputEmailSignupSubmit spec take in parameters that define the email, username, and password allow us to re-use the spec function to address different test conditions we are exploring. Then we take the parameterized function and use it in two different suites with different expected outcomes.

File: src/Tester.js (hunk 2 above)
Comment (TGPSKI) [inline src/Tester.js]: Addressed here: 

https://github.com/pixielabs/cavy/pull/21

File: src/Tester.js (hunk 1 above)
Comment (TGPSKI) [inline src/Tester.js]: Will address in the reporting PR to follow https://github.com/pixielabs/cavy/pull/21

**Comment by TGPSKI** [comment]
@jalada: Breaking down this PR into smaller ones. The first is here: https://github.com/pixielabs/cavy/pull/21

This PR handles adding new Tester.js props, as well as refactoring run() in TestScope.js. This PR does not include any reporting features, and the sample project has not been modified.

**Comment by TGPSKI** [comment]
Closing w/ #21 and #22

**Pull request closed without merging**